lib/src/rewrite.rs

```diff
@@ -69,28 +70,45 @@ pub async fn merge_commit_trees_no_resolve_without_repo(
     index: &dyn Index,
     commits: &[Commit],
 ) -> BackendResult<MergedTree> {
-    if commits.is_empty() {
-        Ok(store
-            .get_root_tree_async(&store.empty_merged_tree_id())
-            .await?)
+    let commit_ids = commits
+        .iter()
+        .map(|commit| commit.id().clone())
+        .collect_vec();
+    let commit_id_merge = find_recursive_merge_commits(store, index, &commit_ids)?;
+    let tree_merge = commit_id_merge
+        .try_map_async(async |commit_id| {
+            let commit = store.get_commit_async(commit_id).await?;
+            let tree = commit.tree_async().await?;
+            Ok::<_, BackendError>(tree.take())
+        })
+        .await?;
+    Ok(MergedTree::new(tree_merge.flatten().simplify()))
+}
+
+/// Find the commits to use as input to the recursive merge algorithm.
+#[instrument(skip(index))]
```

Comment:
nit: Delete instrument? I'm not sure if tracing wrapper was supposed to be inserted to every recursion.